server/s3_put_multiobject_copy_action.h

```diff
@@ -0,0 +1,220 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#pragma once
+
+#ifndef __S3_SERVER_S3_PUT_MULTIPART_COPY_ACTION_H__
+#define __S3_SERVER_S3_PUT_MULTIPART_COPY_ACTION_H__
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include <gtest/gtest_prod.h>
+
+#include "s3_put_object_action_base.h"
+#include "s3_object_data_copier.h"
+#include "s3_object_metadata.h"
+#include "s3_async_buffer.h"
+#include "s3_bucket_metadata.h"
+#include "s3_motr_writer.h"
+#include "s3_part_metadata.h"
+#include "s3_probable_delete_record.h"
+#include "s3_timer.h"
+
+enum class S3CopyPartActionState {
+  empty = 0,         // Initial state
+  validationFailed,  // Any validations failed for request, including metadata
+  probableEntryRecordFailed,
+  newObjOidCreated,         // New object created
+  newObjOidCreationFailed,  // New object create failed
+  writeComplete,            // data write to object completed successfully
+  writeFailed,              // data write to object failed
+  md5ValidationFailed,      // md5 check failed
+  metadataSaved,            // metadata saved for new object
+  metadataSaveFailed,       // metadata saved for new object
+  completed,                // All stages done completely
+};
+
+class S3ObjectDataCopier;
+
+class S3PutMultipartCopyAction : public S3PutObjectActionBase {
+  std::string auth_acl;
+  S3CopyPartActionState s3_copy_part_action_state;
+  std::shared_ptr<S3MotrReaderFactory> motr_reader_factory;
+  std::unique_ptr<S3ObjectDataCopier> object_data_copier;
+  std::shared_ptr<S3ObjectDataCopier> fragment_data_copier;
+  std::vector<std::shared_ptr<S3ObjectDataCopier>> fragment_data_copier_list;
+  std::shared_ptr<S3PartMetadata> part_metadata = NULL;
+  std::shared_ptr<S3ObjectMetadata> object_multipart_metadata = NULL;
+  std::shared_ptr<S3MotrKVSWriterFactory> motr_kv_writer_factory;
+  std::unique_ptr<S3ProbableDeleteRecord> new_probable_del_rec;
+
+  std::shared_ptr<S3ObjectMultipartMetadataFactory> object_mp_metadata_factory;
+  std::shared_ptr<S3PartMetadataFactory> part_metadata_factory;
+
+  bool response_started = false;
+  bool auth_in_progress = false;
+  bool auth_failed = false;
+  bool auth_completed = false;
+    
+  bool motr_write_in_progress = false;
+  bool motr_write_completed = false;  // full object write
+  bool write_failed = false;
+
+  int total_parts_fragment_to_be_copied = 0;
+  int parts_fragment_copied_or_failed = 0;
+  int max_parallel_copy = 0;
+  int parts_frg_copy_in_flight_copied_or_failed = 0;
+  int parts_frg_copy_in_flight = 0;
+  int part_number;
+  std::string upload_id;
+  unsigned motr_write_payload_size;
+  bool if_source_and_destination_same();
+  S3Timer s3_timer;
+
+  int get_part_number() {
```

Comment:
make sure we have enough validation against this part number before consuming, as its coming in request. Check CWE-190 security issue flagged by codacy at line 95.

Also why this function in header file ?

Reply:
> Also why this function in header file ?
It's a one-line function.

Reply:
> make sure we have enough validation against this part number before consuming, as its coming in request. Check CWE-190 security issue flagged by codacy at line 95.

Replaced atoi() with strtol() to avoid CWE-190